homework/list.js

```diff
@@ -52,28 +44,41 @@
       buff.appendChild(li);
     });
     ul.appendChild(buff);
-    _wrapper.appendChild(ul);
-  }
+    this._wrapper.appendChild(ul);
+  },
 
-  function fetchList(afterFetch) {
-    var xhr = new XMLHttpRequest();
-    xhr.open('GET', 'http://127.0.0.1:8000/demo-list-notes.json', true);
-    xhr.responseType = 'json';
-    xhr.onreadystatechange = function(e) {
-      // Watch out: we have a mysterious unknown 'this'.
-      if (this.readyState === 4 && this.status === 200) {
-        var listData = this.response;
-        // The flow ends here.
-        afterFetch(listData);
-      } else if (this.status !== 200 ){
-        // Ignore error in this case.
-      }
+  fetchList() {
+    return new Promise(function(resolve, reject) {
+		var xhr = new XMLHttpRequest();
+        xhr.open('GET', 'http://127.0.0.1:8000/demo-list-notes.json', true);
+        xhr.responseType = 'json';
+        xhr.onreadystatechange = (function(e) {
+          // Watch out: we have a mysterious unknown 'this'.
+          if (this.readyState === 4 && this.status === 200) {
+            var listData = this.response;
+            // The flow ends here.
+            afterFetch(listData);
+          } else if (this.status !== 200 ){
+            // Ignore error in this case.
+          }
     };
     xhr.send();
-  }
-
-  document.addEventListener('DOMContentLoaded', function(event) {
-    start();
   });
+  },
+  
+  start() {
+      this.fetchList().then((function(data){
+        this.updateList(data);
+        this.drawList();
+        this.preloadFirstNote();
+      }).bind(this)).catch((function() {
```

Comment:
**WRONG**: the similar flow on my slides you may directly refered to **is not the whole story**. To catch an error in a Promise without **re-throwing** it, the following chain will not be interrupted if there is any error. And that will cause some unexpected behaviors which now are worse since you also didn't **log** it. The correct way is you firstly log the error to prevent the following chain might from others ignore it incorrectly just as what you're doing, and then you re-throw it to interrupt the whole chain. If you still have any question, **ask me**, just as I always asked you all to do in the class.